### PR TITLE
Hardening receive_all

### PR DESCRIPTION
**Pull request by SergiySW**
- check minimum confirmation threshold
- check winner to be required account head
- don't receive possible successors of confirmed account head

### CODE REVIEW

File: rai/node/wallet.cpp [hunk 1]
```diff
@@ -1228,10 +1228,18 @@ class search_action : public std::enable_shared_from_this<search_action>
...
 						wallet->node.background ([this_l, account, block_l] {
+							rai::block_hash hash (block_l->hash ());
 							rai::transaction transaction (this_l->wallet->node.store.environment, nullptr, true);
-							this_l->wallet->node.active.start (transaction, block_l, [this_l, account](std::shared_ptr<rai::block>, bool) {
+							this_l->wallet->node.active.start (transaction, block_l, [this_l, account, hash](std::shared_ptr<rai::block> winner_l, bool confirmed) {
 								// If there were any forks for this account they've been rolled back and we can receive anything remaining from this account
-								this_l->receive_all (account);
+								if (confirmed && winner_l->hash () == hash)
+								{
+									this_l->receive_all (account, hash);
```
Comment (PlasmaPower) [inline rai/node/wallet.cpp]: Why check `winner_l->hash () == hash`? Why not just check `confirmed`, and then call `this_l->receive_all (account, winner_l->hash ());`? Other than that LGTM.

Comment (SergiySW) [reply rai/node/wallet.cpp]: Yeah, seems better

Comment (SergiySW) [reply rai/node/wallet.cpp]: I guess in rare conditions block_processor probably don't process new winner before callback (as there is queue in block_processor). For this this theoretic possibility I add new extra check for winner existance

**Comment by SergiySW** [comment]
Obsolete

**Pull request closed without merging**